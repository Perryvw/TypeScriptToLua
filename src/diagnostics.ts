import * as ts from "typescript";
import { TranspileError } from "./TranspileError";

export const transpileError = (error: TranspileError): ts.Diagnostic => ({
    file: error.node.getSourceFile(),
    start: error.node.getStart(),
    length: error.node.getWidth(),
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: error.message,
});

export const tstlOptionsAreMovingToTheTstlObject = (tstl: Record<string, any>): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Warning,
    code: 0,
    source: "typescript-to-lua",
    messageText:
        'TSTL options are moving to the "tstl" object. Adjust your tsconfig to look like\n' +
        `"tstl": ${JSON.stringify(tstl, undefined, 4)}`,
});

export const toLoadTransformerItShouldBeTranspiled = (transform: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `To load "${transform}" transformer it should be transpiled or "ts-node" should be installed`,
});

export const couldNotResolveTransformerFrom = (transform: string, base: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `Could not resolve "${transform}" transformer from "${base}".`,
});

export const transformerShouldHaveAExport = (transform: string, importName: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `"${transform}" transformer should have a "${importName}" export`,
});

export const transformerShouldBeATsTransformerFactory = (transform: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `"${transform}" transformer should be a ts.TransformerFactory or an object with ts.TransformerFactory values`,
});

export const watchErrorSummary = (errorCount: number): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Message,
    code: errorCount === 1 ? 6193 : 6194,
    messageText:
        errorCount === 1
            ? "Found 1 error. Watching for file changes."
            : `Found ${errorCount} errors. Watching for file changes.`,
});

const createCommandLineError = <Args extends any[]>(code: number, getMessage: (...args: Args) => string) => (
    ...args: Args
): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code,
    messageText: getMessage(...args),
});

export const unknownCompilerOption = createCommandLineError(
    5023,
    (name: string) => `Unknown compiler option '${name}'.`
);

export const compilerOptionRequiresAValueOfType = createCommandLineError(
    5024,
    (name: string, type: string) => `Compiler option '${name}' requires a value of type ${type}.`
);

export const optionProjectCannotBeMixedWithSourceFilesOnACommandLine = createCommandLineError(
    5042,
    () => "Option 'project' cannot be mixed with source files on a command line."
);

export const cannotFindATsconfigJsonAtTheSpecifiedDirectory = createCommandLineError(
    5057,
    (dir: string) => `Cannot find a tsconfig.json file at the specified directory: '${dir}'.`
);

export const theSpecifiedPathDoesNotExist = createCommandLineError(
    5058,
    (dir: string) => `The specified path does not exist: '${dir}'.`
);

export const compilerOptionExpectsAnArgument = createCommandLineError(
    6044,
    (name: string) => `Compiler option '${name}' expects an argument.`
);

export const argumentForOptionMustBe = createCommandLineError(
    6046,
    (name: string, values: string) => `Argument for '${name}' option must be: ${values}.`
);

export const optionCanOnlyBeSpecifiedInTsconfigJsonFile = createCommandLineError(
    6064,
    (name: string) => `Option '${name}' can only be specified in 'tsconfig.json' file.`
);

export const optionBuildMustBeFirstCommandLineArgument = createCommandLineError(
    6369,
    () => "Option '--build' must be the first command line argument."
);
