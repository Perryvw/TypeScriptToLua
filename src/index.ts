import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileWithSystem } from "./CommandLineParser";
import { CompilerOptions } from "./CompilerOptions";
import { getTranspileOutput, TranspilationResult, TranspiledFile } from "./Transpile";

export { parseCommandLine, ParsedCommandLine, updateParsedConfigFile } from "./CommandLineParser";
export { CompilerOptions, LuaLibImportKind, LuaTarget } from "./CompilerOptions";
export * from "./Emit";
export * from "./LuaAST";
export { LuaLibFeature } from "./LuaLib";
export { LuaPrinter } from "./LuaPrinter";
export { LuaTransformer } from "./LuaTransformer";
export * from "./Transpile";

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {}
): TranspilationResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { transpiledFiles, diagnostics: [...allDiagnostics] };
}

export function transpileProject(fileName: string, options?: CompilerOptions): TranspilationResult {
    const parseResult = parseConfigFileWithSystem(fileName, options);
    if (parseResult.errors.length > 0) {
        return { diagnostics: parseResult.errors, transpiledFiles: new Map() };
    }

    return transpileFiles(parseResult.fileNames, parseResult.options);
}

const libCache: { [key: string]: ts.SourceFile } = {};

/** @internal */
export function createVirtualProgram(
    input: Record<string, string>,
    options?: CompilerOptions
): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: () => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile: () => {},

        getSourceFile: filename => {
            if (filename in input) {
                return ts.createSourceFile(
                    filename,
                    input[filename],
                    ts.ScriptTarget.Latest,
                    false
                );
            }

            if (filename.startsWith("lib.")) {
                if (libCache[filename]) return libCache[filename];
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                const filePath = path.join(typeScriptDir, filename);
                const content = fs.readFileSync(filePath, "utf8");

                libCache[filename] = ts.createSourceFile(
                    filename,
                    content,
                    ts.ScriptTarget.Latest,
                    false
                );

                return libCache[filename];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface VirtualProgramResult {
    file: TranspiledFile;
    diagnostics: ts.Diagnostic[];
}

export function transpileString(
    content: string,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const program = createVirtualProgram({ "main.ts": content }, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get("main.ts"), diagnostics: [...allDiagnostics] };
}

export function transpileVirtualProgram(
    files: Record<string, string>,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const mainFileName = Object.keys(files).find(x => /\bmain\.[a-z]+$/.test(x));
    if (mainFileName === undefined) {
        throw new Error('Virtual program should have a file named "main"');
    }

    const program = createVirtualProgram(files, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get(mainFileName), diagnostics: [...allDiagnostics] };
}
