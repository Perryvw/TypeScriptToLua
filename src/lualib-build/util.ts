import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
