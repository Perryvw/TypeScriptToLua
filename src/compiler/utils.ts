import { create as createResolve } from "enhanced-resolve";
import * as ts from "typescript";
// TODO: Don't depend on CLI?
import * as cliDiagnostics from "../cli/diagnostics";
import { assert } from "../utils";
import * as diagnosticFactories from "./diagnostics";

// https://github.com/webpack/enhanced-resolve/blob/0001f80dacf033ac4a0e690b2766e0965c458266/lib/Resolver.js#L280-L288
export const isResolveError = (error: unknown): error is Error & { details: string } =>
    error instanceof Error && "details" in error;

const resolveConfigImport = createResolve.sync({ extensions: [".js", ".ts", ".tsx"] });

export function loadConfigImport(
    kind: string,
    optionName: string,
    basedir: string,
    query: string,
    importName = "default"
): { error?: ts.Diagnostic; result?: unknown } {
    if (typeof query !== "string") {
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "string") };
    }

    let resolved: string;
    try {
        const result = resolveConfigImport({}, basedir, query);
        assert(typeof result === "string");
        resolved = result;
    } catch (error) {
        if (!isResolveError(error)) throw error;
        return { error: diagnosticFactories.couldNotResolveFrom(kind, query, basedir) };
    }

    const hasNoRequireHook = require.extensions[".ts"] === undefined;
    if (hasNoRequireHook && (resolved.endsWith(".ts") || resolved.endsWith(".tsx"))) {
        try {
            require("ts-node/register/transpile-only");
        } catch (err) {
            if (err.code !== "MODULE_NOT_FOUND") throw err;
            return { error: diagnosticFactories.toLoadItShouldBeTranspiled(kind, query) };
        }
    }

    const commonjsModule = require(resolved);
    const factoryModule = commonjsModule.__esModule ? commonjsModule : { default: commonjsModule };
    const result = factoryModule[importName];
    if (result === undefined) {
        return { error: diagnosticFactories.shouldHaveAExport(kind, query, importName) };
    }

    return { result };
}
