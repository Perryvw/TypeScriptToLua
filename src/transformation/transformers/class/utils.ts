import * as ts from "typescript";
import { DecoratorKind, getCustomDecorators } from "../../utils/decorators";
import { TransformationContext } from "../../context";

export function isStaticNode(node: ts.Node): boolean {
    return node.modifiers !== undefined && node.modifiers.some(m => m.kind === ts.SyntaxKind.StaticKeyword);
}

export function getExtendedTypeNode(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.ExpressionWithTypeArguments | undefined {
    if (node && node.heritageClauses) {
        for (const clause of node.heritageClauses) {
            if (clause.token === ts.SyntaxKind.ExtendsKeyword) {
                const superType = context.checker.getTypeAtLocation(clause.types[0]);
                const decorators = getCustomDecorators(context, superType);
                if (!decorators.has(DecoratorKind.PureAbstract)) {
                    return clause.types[0];
                }
            }
        }
    }
}

export function getExtendedType(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.Type | undefined {
    const extendedTypeNode = getExtendedTypeNode(context, node);
    return extendedTypeNode && context.checker.getTypeAtLocation(extendedTypeNode);
}

export function findInClassOrAncestor(
    context: TransformationContext,
    classDeclaration: ts.ClassLikeDeclarationBase,
    callback: (classDeclaration: ts.ClassLikeDeclarationBase) => boolean
): ts.ClassLikeDeclarationBase | undefined {
    if (callback(classDeclaration)) {
        return classDeclaration;
    }

    const extendsType = getExtendedType(context, classDeclaration);
    if (!extendsType) {
        return undefined;
    }

    const symbol = extendsType.getSymbol();
    if (symbol === undefined) {
        return undefined;
    }

    const symbolDeclarations = symbol.getDeclarations();
    if (symbolDeclarations === undefined) {
        return undefined;
    }

    const declaration = symbolDeclarations.find(ts.isClassLike);
    if (!declaration) {
        return undefined;
    }

    return findInClassOrAncestor(context, declaration, callback);
}
