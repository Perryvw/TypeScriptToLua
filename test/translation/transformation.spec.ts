import * as fs from "fs";
import * as path from "path";
import * as tstl from "../../src";
import { annotationDeprecated } from "../../src/transformation/utils/diagnostics";
import { couldNotResolveRequire } from "../../src/transpilation/diagnostics";
import * as util from "../util";

const targetSpecs: Array<[string, tstl.LuaTarget | undefined, string]> = [
    ["", undefined, "./transformation"],
    ["Luau ", tstl.LuaTarget.Luau, "./transformation/luau"],
];

for (const [name, luaTarget, targetDir] of targetSpecs) {
    const fixturesPath = path.join(__dirname, targetDir);
    const fixtures = fs
        .readdirSync(fixturesPath)
        .filter(f => path.extname(f) === ".ts")
        .sort()
        .map(f => [path.parse(f).name, fs.readFileSync(path.join(fixturesPath, f), "utf8")]);

    test.each(fixtures)(`${name}Transformation (%s)`, (_name, content) => {
        util.testModule(content)
            .setOptions({
                luaLibImport: tstl.LuaLibImportKind.Require,
                luaTarget,
            })
            .ignoreDiagnostics([annotationDeprecated.code, couldNotResolveRequire.code])
            .disableSemanticCheck()
            .expectLuaToMatchSnapshot();
    });
}
