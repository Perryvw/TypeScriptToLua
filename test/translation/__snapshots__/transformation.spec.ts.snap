// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transformation (callNamespace) 1`] = `"Namespace:myFunction();"`;

exports[`Transformation (characterEscapeSequence) 1`] = `
"local quoteInDoubleQuotes = \\"\\\\' \\\\' \\\\'\\";
local quoteInTemplateString = \\"\\\\' \\\\' \\\\'\\";
local doubleQuoteInQuotes = \\"\\\\\\" \\\\\\" \\\\\\"\\";
local doubleQuoteInDoubleQuotes = \\"\\\\\\" \\\\\\" \\\\\\"\\";
local doubleQuoteInTemplateString = \\"\\\\\\" \\\\\\" \\\\\\"\\";
local backQuoteInQuotes = \\"\` \` \`\\";
local backQuoteInDoubleQuotes = \\"\` \` \`\\";
local backQuoteInTemplateString = \\"\` \` \`\\";
local escapedCharsInQuotes = \\"\\\\\\\\ \\\\0 \\\\b \\\\t \\\\n \\\\v \\\\f \\\\\\" \\\\' \`\\";
local escapedCharsInDoubleQUotes = \\"\\\\\\\\ \\\\0 \\\\b \\\\t \\\\n \\\\v \\\\f \\\\\\" \\\\'\\";
local escapedCharsInTemplateString = \\"\\\\\\\\ \\\\0 \\\\b \\\\t \\\\n \\\\v \\\\f \\\\\\" \\\\' \`\\";
local nonEmptyTemplateString = \\"Level 0: \\\\n\\\\t \\" .. (tostring(\\"Level 1: \\\\n\\\\t\\\\t \\" .. (tostring(\\"Level 3: \\\\n\\\\t\\\\t\\\\t \\" .. (tostring(\\"Last level \\\\n --\\") .. \\" \\\\n --\\")) .. \\" \\\\n --\\")) .. \\" \\\\n --\\");"
`;

exports[`Transformation (classExtension1) 1`] = `
"MyClass.myFunction = function(self)
end;"
`;

exports[`Transformation (classExtension2) 1`] = `
"TestClass.myFunction = function(self)
end;"
`;

exports[`Transformation (classExtension3) 1`] = `
"RenamedTestClass.myFunction = function(self)
end;
RenamedMyClass.myFunction = function(self)
end;"
`;

exports[`Transformation (classExtension4) 1`] = `
"MyClass.test = \\"test\\";
MyClass.testP = \\"testP\\";
MyClass.myFunction = function(self)
end;"
`;

exports[`Transformation (classPureAbstract) 1`] = `
"ClassB = ClassB or {};
ClassB.__index = ClassB;
ClassB.prototype = ClassB.prototype or {};
ClassB.prototype.__index = ClassB.prototype;
ClassB.prototype.constructor = ClassB;
ClassB.new = function(...)
    local self = setmetatable({}, ClassB.prototype);
    self:____constructor(...);
    return self;
end;
ClassB.prototype.____constructor = function(self)
end;"
`;

exports[`Transformation (continue) 1`] = `
"do
    local i = 0;
    while i < 10 do
        do
            if i < 5 then
                goto __continue1;
            end
        end
        ::__continue1::
        i = i + 1;
    end
end"
`;

exports[`Transformation (continueConcurrent) 1`] = `
"do
    local i = 0;
    while i < 10 do
        do
            if i < 5 then
                goto __continue1;
            end
            if i == 7 then
                goto __continue1;
            end
        end
        ::__continue1::
        i = i + 1;
    end
end"
`;

exports[`Transformation (continueNested) 1`] = `
"do
    local i = 0;
    while i < 5 do
        do
            if (i % 2) == 0 then
                goto __continue1;
            end
            do
                local j = 0;
                while j < 2 do
                    do
                        if j == 1 then
                            goto __continue3;
                        end
                    end
                    ::__continue3::
                    j = j + 1;
                end
            end
        end
        ::__continue1::
        i = i + 1;
    end
end"
`;

exports[`Transformation (continueNestedConcurrent) 1`] = `
"do
    local i = 0;
    while i < 5 do
        do
            if (i % 2) == 0 then
                goto __continue1;
            end
            do
                local j = 0;
                while j < 2 do
                    do
                        if j == 1 then
                            goto __continue3;
                        end
                    end
                    ::__continue3::
                    j = j + 1;
                end
            end
            if i == 4 then
                goto __continue1;
            end
        end
        ::__continue1::
        i = i + 1;
    end
end"
`;

exports[`Transformation (do) 1`] = `
"local e = 10;
repeat
    do
        e = e - 1;
    end
    ::__continue1::
until not (e > 0);"
`;

exports[`Transformation (enum) 1`] = `
"TestEnum = {};
TestEnum.val1 = 0;
TestEnum[0] = \\"val1\\";
TestEnum.val2 = 2;
TestEnum[2] = \\"val2\\";
TestEnum.val3 = 3;
TestEnum[3] = \\"val3\\";"
`;

exports[`Transformation (enumHeterogeneous) 1`] = `
"TestEnum = {};
TestEnum.val1 = 0;
TestEnum[0] = \\"val1\\";
TestEnum.val2 = 3;
TestEnum[3] = \\"val2\\";
TestEnum.val3 = \\"baz\\";
TestEnum.baz = \\"val3\\";"
`;

exports[`Transformation (enumMembersOnly) 1`] = `
"val1 = 0;
val2 = 2;
val3 = 3;
val4 = \\"bye\\";
local a = val1;"
`;

exports[`Transformation (enumString) 1`] = `
"TestEnum = {};
TestEnum.val1 = \\"foo\\";
TestEnum.foo = \\"val1\\";
TestEnum.val2 = \\"bar\\";
TestEnum.bar = \\"val2\\";
TestEnum.val3 = \\"baz\\";
TestEnum.baz = \\"val3\\";"
`;

exports[`Transformation (exportStatement) 1`] = `
"local exports = exports or {};
local xyz = 4;
exports.xyz = xyz;
exports.uwv = xyz;
do
    local __TSTL_export = require(\\"xyz\\");
    for ____exportKey, ____exportValue in pairs(__TSTL_export) do
        exports[____exportKey] = ____exportValue;
    end
end
do
    local __TSTL_xyz = require(\\"xyz\\");
    local abc = __TSTL_xyz.abc;
    local def = __TSTL_xyz.def;
    exports.abc = abc;
    exports.def = def;
end
do
    local __TSTL_xyz = require(\\"xyz\\");
    local def = __TSTL_xyz.abc;
    exports.def = def;
end
return exports;"
`;

exports[`Transformation (for) 1`] = `
"do
    local i = 1;
    while i <= 100 do
        do
        end
        ::__continue1::
        i = i + 1;
    end
end"
`;

exports[`Transformation (forIn) 1`] = `
"for i in pairs({a = 1, b = 2, c = 3, d = 4}) do
    do
    end
    ::__continue1::
end"
`;

exports[`Transformation (forOf) 1`] = `
"local ____TS_array = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
for ____TS_index = 1, #____TS_array do
    local i = ____TS_array[____TS_index];
    do
    end
    ::__continue1::
end"
`;

exports[`Transformation (functionRestArguments) 1`] = `
"varargsFunction = function(self, a, ...)
    local b = ({...});
end;"
`;

exports[`Transformation (getSetAccessors) 1`] = `
"require(\\"lualib_bundle\\");
MyClass = MyClass or {};
MyClass.__index = MyClass;
MyClass.prototype = MyClass.prototype or {};
MyClass.prototype.____getters = {};
MyClass.prototype.__index = __TS__Index(MyClass.prototype);
MyClass.prototype.____setters = {};
MyClass.prototype.__newindex = __TS__NewIndex(MyClass.prototype);
MyClass.prototype.constructor = MyClass;
MyClass.new = function(...)
    local self = setmetatable({}, MyClass.prototype);
    self:____constructor(...);
    return self;
end;
MyClass.prototype.____constructor = function(self)
end;
MyClass.prototype.____getters.field = function(self)
    return self._field + 4;
end;
MyClass.prototype.____setters.field = function(self, v)
    self._field = v * 2;
end;
local instance = MyClass.new();
instance.field = 4;
local b = instance.field;
local c = (4 + instance.field) * 3;"
`;

exports[`Transformation (interfaceIndex) 1`] = `
"local a = {};
a.abc = \\"def\\";"
`;

exports[`Transformation (methodRestArguments) 1`] = `
"MyClass = MyClass or {};
MyClass.__index = MyClass;
MyClass.prototype = MyClass.prototype or {};
MyClass.prototype.__index = MyClass.prototype;
MyClass.prototype.constructor = MyClass;
MyClass.new = function(...)
    local self = setmetatable({}, MyClass.prototype);
    self:____constructor(...);
    return self;
end;
MyClass.prototype.____constructor = function(self)
end;
MyClass.prototype.varargsFunction = function(self, a, ...)
    local b = ({...});
end;"
`;

exports[`Transformation (modulesChangedVariableExport) 1`] = `
"local exports = exports or {};
exports.foo = 1;
return exports;"
`;

exports[`Transformation (modulesClassExport) 1`] = `
"local exports = exports or {};
exports.TestClass = exports.TestClass or {};
exports.TestClass.__index = exports.TestClass;
exports.TestClass.prototype = exports.TestClass.prototype or {};
exports.TestClass.prototype.__index = exports.TestClass.prototype;
exports.TestClass.prototype.constructor = exports.TestClass;
exports.TestClass.new = function(...)
    local self = setmetatable({}, exports.TestClass.prototype);
    self:____constructor(...);
    return self;
end;
exports.TestClass.prototype.____constructor = function(self)
end;
return exports;"
`;

exports[`Transformation (modulesClassWithMemberExport) 1`] = `
"local exports = exports or {};
exports.TestClass = exports.TestClass or {};
exports.TestClass.__index = exports.TestClass;
exports.TestClass.prototype = exports.TestClass.prototype or {};
exports.TestClass.prototype.__index = exports.TestClass.prototype;
exports.TestClass.prototype.constructor = exports.TestClass;
exports.TestClass.new = function(...)
    local self = setmetatable({}, exports.TestClass.prototype);
    self:____constructor(...);
    return self;
end;
exports.TestClass.prototype.____constructor = function(self)
end;
exports.TestClass.prototype.memberFunc = function(self)
end;
return exports;"
`;

exports[`Transformation (modulesFunctionExport) 1`] = `
"local exports = exports or {};
exports.publicFunc = function(self)
end;
return exports;"
`;

exports[`Transformation (modulesFunctionNoExport) 1`] = `
"publicFunc = function(self)
end;"
`;

exports[`Transformation (modulesImportAll) 1`] = `"local Test = require(\\"test\\");"`;

exports[`Transformation (modulesImportNamed) 1`] = `
"local __TSTL_test = require(\\"test\\");
local TestClass = __TSTL_test.TestClass;"
`;

exports[`Transformation (modulesImportNamedSpecialChars) 1`] = `
"local __TSTL_kebab_module = require(\\"kebab-module\\");
local TestClass = __TSTL_kebab_module.TestClass;
local __TSTL_dollar_module = require(\\"dollar$module\\");
local TestClass = __TSTL_dollar_module.TestClass;
local __TSTL_singlequote_module = require(\\"singlequote'module\\");
local TestClass = __TSTL_singlequote_module.TestClass;
local __TSTL_hash_module = require(\\"hash#module\\");
local TestClass = __TSTL_hash_module.TestClass;
local __TSTL_space_module = require(\\"space module\\");
local TestClass = __TSTL_space_module.TestClass;"
`;

exports[`Transformation (modulesImportRenamed) 1`] = `
"local __TSTL_test = require(\\"test\\");
local RenamedClass = __TSTL_test.TestClass;"
`;

exports[`Transformation (modulesImportRenamedSpecialChars) 1`] = `
"local __TSTL_kebab_module = require(\\"kebab-module\\");
local RenamedClass = __TSTL_kebab_module.TestClass;
local __TSTL_dollar_module = require(\\"dollar$module\\");
local RenamedClass = __TSTL_dollar_module.TestClass;
local __TSTL_singlequote_module = require(\\"singlequote'module\\");
local RenamedClass = __TSTL_singlequote_module.TestClass;
local __TSTL_hash_module = require(\\"hash#module\\");
local RenamedClass = __TSTL_hash_module.TestClass;
local __TSTL_space_module = require(\\"space module\\");
local RenamedClass = __TSTL_space_module.TestClass;"
`;

exports[`Transformation (modulesImportWithoutFromClause) 1`] = `"require(\\"test\\");"`;

exports[`Transformation (modulesNamespaceExport) 1`] = `
"local exports = exports or {};
exports.TestSpace = exports.TestSpace or {};
local TestSpace = exports.TestSpace;
do
end
return exports;"
`;

exports[`Transformation (modulesNamespaceExportEnum) 1`] = `
"local exports = exports or {};
exports.test = exports.test or {};
local test = exports.test;
do
    test.TestEnum = {};
    test.TestEnum.foo = \\"foo\\";
    test.TestEnum.foo = \\"foo\\";
    test.TestEnum.bar = \\"bar\\";
    test.TestEnum.bar = \\"bar\\";
end
return exports;"
`;

exports[`Transformation (modulesNamespaceNestedWithMemberExport) 1`] = `
"local exports = exports or {};
exports.TestSpace = exports.TestSpace or {};
local TestSpace = exports.TestSpace;
do
    TestSpace.TestNestedSpace = TestSpace.TestNestedSpace or {};
    local TestNestedSpace = TestSpace.TestNestedSpace;
    do
        TestNestedSpace.innerFunc = function(self)
        end;
    end
end
return exports;"
`;

exports[`Transformation (modulesNamespaceNoExport) 1`] = `
"TestSpace = TestSpace or {};
do
end"
`;

exports[`Transformation (modulesNamespaceWithMemberExport) 1`] = `
"local exports = exports or {};
exports.TestSpace = exports.TestSpace or {};
local TestSpace = exports.TestSpace;
do
    TestSpace.innerFunc = function(self)
    end;
end
return exports;"
`;

exports[`Transformation (modulesNamespaceWithMemberNoExport) 1`] = `
"local exports = exports or {};
exports.TestSpace = exports.TestSpace or {};
local TestSpace = exports.TestSpace;
do
    local innerFunc;
    innerFunc = function(self)
    end;
end
return exports;"
`;

exports[`Transformation (modulesVariableExport) 1`] = `
"local exports = exports or {};
exports.foo = \\"bar\\";
return exports;"
`;

exports[`Transformation (modulesVariableNoExport) 1`] = `"local foo = \\"bar\\";"`;

exports[`Transformation (namespace) 1`] = `
"myNamespace = myNamespace or {};
do
    local nsMember;
    nsMember = function(self)
    end;
end"
`;

exports[`Transformation (namespaceMerge) 1`] = `
"MergedClass = MergedClass or {};
MergedClass.__index = MergedClass;
MergedClass.prototype = MergedClass.prototype or {};
MergedClass.prototype.__index = MergedClass.prototype;
MergedClass.prototype.constructor = MergedClass;
MergedClass.new = function(...)
    local self = setmetatable({}, MergedClass.prototype);
    self:____constructor(...);
    return self;
end;
MergedClass.prototype.____constructor = function(self)
    self.propertyFunc = function(____)
    end;
end;
MergedClass.staticMethodA = function(self)
end;
MergedClass.staticMethodB = function(self)
    self:staticMethodA();
end;
MergedClass.prototype.methodA = function(self)
end;
MergedClass.prototype.methodB = function(self)
    self:methodA();
    self:propertyFunc();
end;
MergedClass = MergedClass or {};
do
    MergedClass.namespaceFunc = function(self)
    end;
end
local mergedClass = MergedClass.new();
mergedClass:methodB();
mergedClass:propertyFunc();
MergedClass:staticMethodB();
MergedClass:namespaceFunc();"
`;

exports[`Transformation (namespaceNested) 1`] = `
"myNamespace = myNamespace or {};
do
    myNamespace.myNestedNamespace = myNamespace.myNestedNamespace or {};
    local myNestedNamespace = myNamespace.myNestedNamespace;
    do
        local nsMember;
        nsMember = function(self)
        end;
    end
end"
`;

exports[`Transformation (namespacePhantom) 1`] = `
"nsMember = function(self)
end;"
`;

exports[`Transformation (returnDefault) 1`] = `
"myFunc = function(self)
    return;
end;"
`;

exports[`Transformation (shorthandPropertyAssignment) 1`] = `
"local f;
f = function(____, x)
    return ({x = x});
end;"
`;

exports[`Transformation (tryCatch) 1`] = `
"do
    local ____TS_try, er = pcall(function()
        local a = 42;
    end);
    if not ____TS_try then
        local b = \\"fail\\";
    end
end"
`;

exports[`Transformation (tryCatchFinally) 1`] = `
"do
    local ____TS_try, er = pcall(function()
        local a = 42;
    end);
    if not ____TS_try then
        local b = \\"fail\\";
    end
    do
        local c = \\"finally\\";
    end
end"
`;

exports[`Transformation (tryFinally) 1`] = `
"do
    pcall(function()
        local a = 42;
    end);
    do
        local b = \\"finally\\";
    end
end"
`;

exports[`Transformation (tupleReturn) 1`] = `
"tupleReturn = function(self)
    return 0, \\"foobar\\";
end;
tupleReturn(_G);
noTupleReturn(_G);
local a, b = tupleReturn(_G);
local c, d = table.unpack(noTupleReturn(_G));
a, b = tupleReturn(_G);
c, d = table.unpack(noTupleReturn(_G));
local e = ({tupleReturn(_G)});
local f = noTupleReturn(_G);
e = ({tupleReturn(_G)});
f = noTupleReturn(_G);
foo(_G, ({tupleReturn(_G)}));
foo(_G, noTupleReturn(_G));
tupleReturnFromVar = function(self)
    local r = {1, \\"baz\\"};
    return table.unpack(r);
end;
tupleReturnForward = function(self)
    return tupleReturn(_G);
end;
tupleNoForward = function(self)
    return ({tupleReturn(_G)});
end;
tupleReturnUnpack = function(self)
    return table.unpack(tupleNoForward(_G));
end;"
`;

exports[`Transformation (typeAssert) 1`] = `
"local test1 = 10;
local test2 = 10;"
`;

exports[`Transformation (while) 1`] = `
"local d = 10;
while d > 0 do
    do
        d = d - 1;
    end
    ::__continue1::
end"
`;
